File: shared_resources/src/main/java/com/thewizrd/shared_resources/SimpleLibrary.java
package com.thewizrd.shared_resources;

import android.content.Context;

public final class SimpleLibrary {
    private ApplicationLib mApp;
    private Context mContext;

    private static SimpleLibrary sSimpleLib;

    public SimpleLibrary() {

    }

    public SimpleLibrary(ApplicationLib app) {
        mApp = app;
        mContext = app.getAppContext();
    }

    public static SimpleLibrary getInstance() {
        if (sSimpleLib == null)
            sSimpleLib = new SimpleLibrary();

        return sSimpleLib;
    }

    public static void init(ApplicationLib app) {
        if (sSimpleLib == null || sSimpleLib.mApp == null || sSimpleLib.mContext == null)
            sSimpleLib = new SimpleLibrary(app);
    }

    public static void unRegister() {
        sSimpleLib = null;
    }

    public ApplicationLib getApp() {
        return mApp;
    }

    public Context getAppContext() {
        return mContext;
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/SimpleLibrary.kt
package com.thewizrd.shared_resources

import android.annotation.SuppressLint
import android.content.Context

class SimpleLibrary private constructor() {
    private var mApp: ApplicationLib? = null
    private var mAppContext: Context? = null

    private constructor(app: ApplicationLib) : this() {
        this.mApp = app
        this.mAppContext = app.appContext
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        @JvmStatic
        private var sSimpleLib: SimpleLibrary? = null

        @JvmStatic
        val instance: SimpleLibrary
            get() {
                if (sSimpleLib == null) {
                    sSimpleLib = SimpleLibrary()
                }

                return sSimpleLib!!
            }

        @JvmStatic
        fun initialize(app: ApplicationLib) {
            if (sSimpleLib == null) {
                sSimpleLib = SimpleLibrary(app)
            } else {
                sSimpleLib!!.mApp = app
                sSimpleLib!!.mAppContext = app.appContext
            }
        }

        fun unregister() {
            sSimpleLib = null
        }
    }

    val app: ApplicationLib
        get() {
            return mApp!!
        }

    val appContext: Context
        get() {
            return mAppContext!!
        }
}